util/overlay/overlay_linux.go

```diff
@@ -429,3 +430,72 @@ func compareFileContent(p1, p2 string) (bool, error) {
 		}
 	}
 }
+
+var (
+	diffSupported     bool
+	diffSupportedOnce sync.Once
+)
+
+func IsDiffSupported() bool {
+	diffSupportedOnce.Do(func() {
+		if err := isDiffSupported(); err != nil {
+			logrus.WithError(err).Warnf("overlayfs differ is not supported")
+			diffSupported = false
+		} else {
+			diffSupported = true
+		}
+	})
+	return diffSupported
+}
+
+func isDiffSupported() error {
```

Comment:
And check user namespace: https://github.com/torvalds/linux/commit/2d2f2d7322ff43e0fe92bf8cccdc0b09449bf2e1

> Disable redirect_dir and metacopy options, because these would allow
privilege escalation through direct manipulation of the
"user.overlay.redirect" or "user.overlay.metacopy" xattrs.

Reply:
So we should always append `redirect_dir=off` to the overlayfs mount option (if kernel supports redirect_dir)?

Reply:
> 

s/always/when in UserNS/

Reply:
The behavior with `userxattr` is really weird. The kernel forces `redirect_dir` off when `userxattr` is provided, but it also throws an error if you specify any options for `redirect_dir`, even setting it to `off`: https://github.com/torvalds/linux/blob/cb690f5238d71f543f4ce874aa59237cf53a877c/fs/overlayfs/super.c#L726-L745

So I think we should skip appending any `redirect_dir` option when in a user ns to get the behavior we want.

Reply:
Using `nofollow`  ([redirects are not created and not followed](https://www.kernel.org/doc/Documentation/filesystems/overlayfs.txt)) didn't return errors.

Based on the discussion with @tonistiigi (https://github.com/moby/buildkit/pull/2491#pullrequestreview-818484141), we should always append `redirect_dir=nofollow` when `/sys/module/overlay/parameters/redirect_dir` exists?

> If no error appears then I'd rather use a mount option that always disables this and overwrites global default until differ can handle it.

Reply:
> Using nofollow (redirects are not created and not followed) didn't return errors.

Oh cool, good catch. The only downside to `nofollow` is that pre-existing snapshots created while the kernel was defaulting `redirect_dir=on` will no longer appear correctly (I don't think so anyways, haven't tested). This is only an issue if a user is already running on a kernel with that default and then upgrades to a commit from the master branch, so it's probably not a huge deal until the next release. But if it's easy to skip appending any `redirect_dir=` option at all when `userxattr` is present that might be slightly safer imo.

Reply:
> But if it's easy to skip appending any redirect_dir= option at all when userxattr is present that might be slightly safer imo.

@AkihiroSuda SGTY?

Reply:
Thank you for the review. Fixed the PR based on the comments:

- Disable redirect_dir using the mount option `redirect_dir=off` when `/sys/module/overlay/parameters/redirect_dir` exists.
- if buildkit is in an user namespace, we disable redirect_dir using `redirect_dir=nofollow` (using `off` results in an error).